clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -1075,7 +1110,37 @@ public void onFailure(RuntimeException e) {
 
     }
 
-    protected static class Generation {
+    protected void loadPreviousGenerationFromLocal(String clientId) {
+        File generationFile = new File(generationDir, clientId);
+        if (!generationFile.exists()) {
+            return;
+        }
+        try {
+            ObjectInputStream generationIn = new ObjectInputStream(new FileInputStream(generationFile));
+            generation = (Generation) generationIn.readObject();
+            log.info("Successfully loaded previous generation with client id: {}", clientId);
+            if (!generationFile.delete()) {
+                log.error("generation file delete failed.");
+            }
+            generationIn.close();
+        } catch (Exception e) {
+            log.error(String.format("Exception when reading generation file at [%s]: [%s]", generationFile.getPath(), e.toString()));
+        }
+    }
+
+    protected void saveGenerationToLocal(String clientId, Generation generation) {
+        File generationFile = new File(generationDir, clientId);
```

Comment:
Should we check if generationDir exists first in case it was deleted?

Reply:
I think this should be fine, since the generationDir is created upon coordinator start.